File: Problems/Problems/Leetcode/C++/searchMatrix.cpp
class Solution {
public:
    int binarySearch(vector<int>& arr, int l, int r, int x) {
        if (r >= l) {
            int mid = l + (r - l) / 2;
            if (arr[mid] == x)
                return mid;
            if (arr[mid] > x)
                return binarySearch(arr, l, mid - 1, x);
            return binarySearch(arr, mid + 1, r, x);
        }
        return -1;
    }
    
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        for(int i = 0; i < matrix.size(); i++) {
            if (!(target > matrix[i][matrix[i].size()-1])) {
                int a = binarySearch(matrix[i], 0, matrix[i].size()-1, target);
                return a != -1;
            }
        }
        return false;
    }
};

File: Problems/Problems/Leetcode/Python/searchMatrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def binarySearch(arr, l, r, x):
            # Check base case
            if r >= l:
                mid = l + (r - l) // 2

                # If element is present at the middle itself
                if arr[mid] == x:
                    return mid

                # If element is smaller than mid, then it
                # can only be present in left subarray
                elif arr[mid] > x:
                    return binarySearch(arr, l, mid-1, x)

                # Else the element can only be present
                # in right subarray
                else:
                    return binarySearch(arr, mid + 1, r, x)

            else:
                # Element is not present in the array
                return -1
        for i in range(len(matrix)):
            if not target > matrix[i][-1]:
                a = binarySearch(matrix[i], 0, len(matrix[i])-1, target)
                return a != -1
        return False
        